README.md

```diff
@@ -0,0 +1,219 @@
+     in the app.
+
+## Shopping
+
+Let's say we want to write an e-commerce site. Users can come to the site, add some items to their shopping cart, and then pay for all of them at the end. Simple.
+
+Where does the user's shopping cart live?
+
+We could create a `Cart` model, which `has_and_belongs_to_many` `Item`s. But that leaves a fundamental problem: how do we know what cart to load when we get a request?
+
+Remember what's included in an HTTP request:
+  * an [HTTP verb][rfc_http_methods], like `GET`, `PUT`, or `POST`
+  * a path
+  * various headers
+
+HTTP servers are typically stateless. Your Rails server has access to a database, but the request needs to have information about what database rows to look up.
+
+For example, `GET` requests usually encode this information in the path. When you write a route matching `/items/:item_id`, you are telling Rails to pull the value `id` from the request path and save it in `params[:id]`. In your `items_controller`, you'll probably have a method that looks something like:
+
+    def show
+      @item = Item.find!(params[:item_id])
+    end
+
+Which loads the row for that item from the database and returns it as an ActiveRecord model object, which your `show.html.erb` then renders.
+
+If we want to be able to retrieve the current cart, we need to its id to be somewhere in the HTTP request. Specifically, it must be in the path or the headers.
+
+It would be possible, though quite convoluted, to store this information in the path. This would have strange effects: since the path is shown in the browser's URL bar, a user who copies a URL and sends it to a friend ("check out this neat skirt!") would also be copying their shopping cart ID. Upon loading the page, the friend would
+see what's in the user's cart. That is probably not what we want.
+
+Cookies allow us to store this information in the only other place available to us: HTTP headers.
+
+## What's a cookie, anyway?
+
+Let's see what [the spec][rfc_cookies] has to say:
+
+      This section outlines a way for an origin server to send state
+      information to a user agent and for the user agent to return the
+      state information to the origin server.
+
+      To store state, the origin server includes a Set-Cookie header in an
+      HTTP response.  In subsequent requests, the user agent returns a
+      Cookie request header to the origin server.  The Cookie header
+      contains cookies the user agent received in previous Set-Cookie
+      headers.  The origin server is free to ignore the Cookie header or
+      use its contents for an application-defined purpose.
+
+Their example is quite simple:
+
+== Server -> User Agent ==
+      Set-Cookie: SID=31d4d96e407aad42
+
+      == User Agent -> Server ==
+      Cookie: SID=31d4d96e407aad42
+
+In this example, the server responds to a request with the `Set-Cookie` header. This header sets the value of the `SID` cookie to `31d4d96e407aad42`.
```

Comment:
What would be even cooler is to sort of teach students how to read that technical specs. Maybe have a sentence of the spec and then "translate" it into something a bit more english-y?